### ISSUE TITLE
Compiling * .h files from idl

### ISSUE THREAD

**Issue by th3g3ntl3m3m** [eclipse-cyclonedds/cyclonedds#131]
Please help me figure out how to create my own types from idl files. I tried to compile a modified HelloWorldType example, I didn’t get an error: 
> Could not find or load main class org.eclipse.cyclonedds.compilers.Idlс

I use VS Studio 2015

**Comment by k0ekk0ek** [comment]
Hi @th3g3ntl3m3m. A script named "dds_idlc.bat" (on Windows) should be generated that allows you to easily use the IDL compiler. It is located under the *idlc* directory in the *build* directory you've chosen. That script should automatically set the classpath before starting the Java compiler. I suspect that'll solve your problem. If it doesn't, please list the steps you've used to build Cyclone DDS and how you intend to call the compiler.

**Comment by th3g3ntl3m3m** [comment]
But could you describe in more detail how the generation process from idl to * .h should take place?

**Comment by k0ekk0ek** [comment]
I can certainly try. When you execute the script, either dds_idlc on Linux or dds_idlc.bat on Windows, it runs a Java program that takes an IDL file. The Java program basically creates an in memory representation of the IDL file that is easy to traverse and calls a backend that walks over the representation and spits out C source code. Basically like every compiler works. Don't know if that answers your question, if it doesn't, please elaborate a bit on what you're trying to achieve.

**Comment by th3g3ntl3m3m** [comment]
Thanks, I figured out the compilation, but I would like to clarify a few questions:
1) Is this library compatible with ROS2?
2) and how the type structure is formed
3) what types are available to create your message
I created my idl file. I would like to clarify why the keylist is needed?
`module DiagnosisFuctionalySystem`
`{`
`  struct Msg`
`  {`
`	string name;`
`	string status;`
`	string status_description;`
`	boolean pulse;`
`	string params;	 `
`  };`
`  //#pragma keylist name status status_description pulse params`
`};`

**Comment by eboasson** [comment]
Hi @th3g3ntl3m3m ,

I don't known the innards of this IDL compiler, but I can help with you with 1 & 3:

> 1. Is this library compatible with ROS2?

Not yet ... this is one of the big things that I hope to see addressed in the near future, as Cyclone would be a very good (the aim is the best 😁) DDS implementation for ROS2. Quite some time ago, I built a partial RMW layer for running ROS2 on Cyclone (see https://github.com/atolab/rmw_cyclonedds) but ran into some problems implementing the introspection features of ROS2. The necessary work in Cyclone has since been done, but I haven't had time to go back and complete the RMW layer.

That code relies on some implementation details of Cyclone that have changed a bit since and it won't build against the current version of Cyclone. This is nothing fundamental, it is just trouble with some definitions being invisible. There is no reason to keep those definitions out of the installed set of header files and neither is there a reason to keep functions private to the library, but it will take some work to address those those issues, even if they are all easily addressed.

Finally, the latest ROS2 release also adds various features, and I haven't studied exactly what the new features mean for ``rmw_cyclonedds``. I'm unfortunately not in a position that I can ask anyone to do that, but I really hope someone with knowledge of ROS2 would lend a hand …

> 3. what types are available to create your message

That's just the traditional IDL types — octet, char, (unsiged) short, (unsigned) long, (unsigned) long long, float, double, string, sequences, arrays, enums, structs and unions. I am pretty sure there are some issues with the preprocessor and serialiser, e.g., I'm pretty sure recursive types are broken. The "simple" stuff works fine.

Work on a new preprocessor with support for xtypes and all the hard cases is under way.

> 3. I would like to clarify why the keylist is needed?

A topic's type definition can recursively refer to many types (I know of systems where the number of types involved is massive), and therefore it makes a lot of sense to only generate the metadata for the types that are actually used as topics.

Part of the topic definition is the list of key fields, so this ``#pragma keylist`` is a very practical way of addressing both things in a very simple way.

IDL4 standardises definining which fields are key fields (before that it was implementation-defined). The new preprocessor being built will accept the standardised way in addition to continuing to accept the ``#pragma keylist``.

**Comment by th3g3ntl3m3m** [comment]
> I don't known the innards of this IDL compiler, but I can help with you with 1 & 3:
> 
> > 1. Is this library compatible with ROS2?
> 
> Not yet ... this is one of the big things that I hope to see addressed in the near future, as Cyclone would be a very good (the aim is the best 😁) DDS implementation for ROS2. Quite some time ago, I built a partial RMW layer for running ROS2 on Cyclone (see https://github.com/atolab/rmw_cyclonedds) but ran into some problems implementing the introspection features of ROS2. The necessary work in Cyclone has since been done, but I haven't had time to go back and complete the RMW layer.
    ---------
My question was a bit different, is it possible at this stage to organize a message exchange between Cyclone DDS and ROS2 using Opencplice DDS?
For example, there are 2 PCs on one installed ROS2, and on the other an application using Cyclone DDS. PCs are in 1 network. And a standard String message such as "Hello world" is sent. Can ROS2 accept the message published by Cyclone, and whether Cyclone will be able to accept the message published by ROS2.

**Comment by eboasson** [comment]
> My question was a bit different, is it possible at this stage to organize a message exchange between Cyclone DDS and ROS2 using Opencplice DDS?

That's a very different question indeed! The answer is simply: yes, that works. The interoperability between different DDS implementations is pretty good, so it doesn't even matter all that much whether you use OpenSplice with ROS2, FastRTPS or RTI Connext.

Obviously, you do need to get the type definition and topic names, &c., correct. I believe the OpenSplice backend for ROS2 generates an IDL file that it feeds into OpenSplice's IDL preprocessor, so you might as well see if you can re-use those generated IDL files. That might save some hassle.

**Comment by i-and** [comment]
Can anyone tell how ROS2 supports the concept of the variability of the types of topics in the process of the life cycle of the system? At what level is the DDS-XTypes standard supported?

**Comment by eboasson** [comment]
@i-and ultimately the best people to answer that question are those developing ROS2, but based on my understanding there is, as of today, no support for type evolution in ROS2. That said, ROS2 recently changed its IDL processing pipeline, moving to using OMG IDL directly, and I would be surpised if it wouldn't continue evolving, first integrating XTypes, and eventually becoming an alternative API for DDS. I'm not sure about the time scale, though, it might be a few years.

I have a number of reasons to hold that opinion — please note that this is just my peronsal opinion — and this morning I fancy climbing on my little soap-box:
* An important driver for ROS2 and its use of DDS was the support of real-time applications, and for this, the original model of ROS was really too limited. At some point I've argued that it had better address the problem addressed by key fields in DDS (there are alternatives, such as subscriptions to topics with wildcards, but those bring a host of other complications). Whether or not that had any effect, I do note that ROS2 now explicitly passes the ``@Key`` IDL4 annotation to the DDS implementation's IDL preprocessor, so something is happening there 😁. It is in my view only a matter of time until the ROS2 API will evolve to recognise these as first-class concepts.
* Based on my (limited) exposure to ROS2-based systems, it appears these systems are typically provisioned with a single software version, and so evolution isn't a big deal yet. However, there's a definite push towards using it in industrial settings (e.g., [ROS-Industrial](https://rosindustrial.org)), and it stands to reason that this will result in requiring interoperability between different versions of software with slightly different interfaces in the long run. The wire interoperability for such cases is pretty well covered by DDS (plenty of examples in the field), but the ROS message type compatibility still needs to be addressed.
The "true" DDS model for evolving the data model really isn't XTypes, but rather adding topics and combining the data at the subscribers (or in dedicated conversion processes), but I've been around for long enough to know that's not going to happen in practice. The reason is that the path of local least resistance is type evolution like XTypes, and that moreover the DDS API completely fails at providing the tools needed to easily and efficiently publish or combine multiple topics. (The MultiTopics are woefully underspecified and the DLRL was a nice idea fraught with problems that eventually got removed from the specification.)
* The systems today are mostly smallish still, and so fault-tolerance and dynamic extensibility isn't a very big deal yet. The launch system of ROS2 recently got a nice upgrade in features, but in the end, with growing systems size and longevity, it will probably morph into a declarative language for stating what components should exist, and what the conditions for mapping the components to hardware are. ([Fog05](https://github.com/eclipse/fog05) is a good example of what I am thinking of.) Now, this part isn't "DDS" per se, but the thing is that the ability to construct, run, monitor, manage, ... such systems involves a few details not-yet covered by ROS2 that have been around in DDS from before DDS even existed ([this](https://books.google.nl/books?id=20ThBwAAQBAJ&lpg=PA53&ots=tqPWg_3d3V&dq=SPLICE%20boasson&pg=PA44#v=onepage&q=SPLICE%20boasson&f=false) was already a little dated at the time of publication but it is in a rather unique book).
The key issue is proper handling of historical data — the "transient" data of DDS [*], rather than the "transient-local" that ROS2 offers. Properly dealing with that requires thinking about the system as having a state, where that state is described by the data published by applications. The most important bit that is missing in ROS2 but mostly present in DDS is dealing with this state, and especially getting rid of it when it is no longer relevant. So I foresee a day that ROS2 will want to add "transient" data, and will see itself forced to import various attendant features of DDS.
(Note that there is and will always be a use for simple (publish-subscribe) messaging such as that offered by ROS2 and DDS in its "volatile" and "transient-local" guise.)

Together, those will push ROS2 towards something very DDS like; the rest of DDS consists of self-evident additions like queries, filters, notifications.

In a way it is a pity that ROS2 abstracted from DDS and now has to re-invent the wheel; but on the other hand, it has proven to be very hard to get people to use DDS to its advantage, whereas ROS2 is attracting people in great numbers. Without those numbers not much will happen — and so it appears that the path chosen by OSRF is at the very least a good one.

(@dirk-thomas, I'd appreciate it if you would correct any blatant misrepresentations. I've had a look at the recent changes, but it is very well possible that I missed something.)

[*] Cyclone DDS doesn't yet support transient data ... of all the things that are suboptimal, wrong or lacking in Cyclone, this is the one that bothers me most ...

**Comment by i-and** [comment]
@eboasson, thank you for your very interesting comments.
